### ISSUE TITLE
Test with Tornado 5.0b1

### ISSUE THREAD

**Issue by bdarnell** [saltstack/salt#45790]
Tornado 5.0 is currently in beta, and I'm filing issues with prominent
users of Tornado to give them a chance to test it before release and
provide feedback on the changes and make adjustments as necessary. You
can install it with `pip install --pre tornado==5.0b1`. I expect to
release the final version of Tornado 5.0 in early March, about a month
away.

Full release notes are
[here](http://www.tornadoweb.org/en/latest/releases/v5.0.0.html). The
major change is that the `IOLoop` is now always backed by `asyncio` on
Python 3, which leads to various subtle changes at runtime but is
mostly compatible with previous versions. A more superficial but
source-incompatible change is that the `io_loop` argument to many
functions in Tornado has been removed and `IOLoop.current()` must be
used instead.

If you've already tested with Tornado 5.0b1 (or temporarily pinned
your dependency to 4.x so you can postpone dealing with this), feel
free to close this issue.

**Comment by garethgreenaway** [comment]
@bdarnell Thanks for the report.  @gtmanfred is on the case.

**Comment by gtmanfred** [comment]
@bdarnell is there a reason that you are just dropping the io_loop all together?  This will break all older versions of salt.

Would it be possible that yall could just swallow the value?  I tested with this, and it seems to work fine for us, the only problem is the error that there is no allowed io_loop variable anymore, which will break older versions.

Also, with not being able to pass in an io_loop, is there any possiblity to run an io_loop inside of an io_loop?

https://github.com/saltstack/salt/blob/2017.7/salt/engines/webhook.py#L84

We use multiprocessing to spin out these extra tasks called "engines".  This one in particular creates a new io_loop for a tiny web/api server, that doesn't really belong on the regular salt io_loop.

Without being able to pass in an io_loop, will it need to be rewritten to handle using IOLoop.current() or is there another way to make a seperate loop?

Thanks,
Daniel

**Comment by bdarnell** [comment]
Accepting the `io_loop` argument but ignoring it would be very confusing if you were passing any other IOLoop than the current one. And if you were passing `IOLoop.current()`, you could simply remove the argument and be compatible with both old and new versions of Tornado.

Being able to run an IOLoop inside an IOLoop is unfortunately something that is lost with the move to asyncio, since asyncio doesn't allow this. If this is something you need, then I'm afraid there are only hacky solutions. The most straightforward is probably to spawn a second thread and create/run the IOLoop there.

**Comment by gtmanfred** [comment]
Ok, we will have to discuss how to best work on fixing this.

You should also let https://github.com/zeromq/pyzmq know, because it looks like this change is going to break some stuff that they are using as well.

**Comment by bdarnell** [comment]
Yes, I've already notified minrk for pyzmq (via https://github.com/jupyter/jupyter/issues/317).

**Comment by cachedout** [comment]
@bdarnell Thanks for reaching out to us. That is really kind.

I can completely understand how you wouldn't want to accept but ignore an `io_loop` argument but might we find a middle ground here? Perhaps logging an error so that it's apparent to a user but it doesn't break entirely?

This puts us in a really tough spot because I don't think we can test and distribute versions of Salt which switch over to using `.current()` everywhere in time for the 5.0 release of Tornado. This, of course, means that there will be a potentially large number of infrastructures that become unmanageable with Salt upon doing a Tornado upgrade, which is most undesirable. :]

We're absolutely going to move to your recommended approach but we were caught by surprise here. Is the Tornado team open to any type of discussion about delaying this change to a later release or is this decision fixed in stone? I'm happy to discuss this offline in more depth if you want but a reply here is of course totally fine as well.

Thanks again for reaching out here. It's really nice to work with libraries that have such proactive maintainers! :]

**Comment by bdarnell** [comment]
How much time would you need? Moving this to a later release is not an option, but the timing of 5.0 is not set in stone. In the meantime could you pin your dependency on `tornado<5.0`? (that's probably a good practice in general - once you've updated for 5.0 you probably want to depend on `<6.0`)

**Comment by cachedout** [comment]
Hi @bdarnell. The issue is really that because it's not pinned to < 5.0 in currently release version of Salt, we have to re-test and re-release multiple versions _and_ we have to give users, some of whom have very large production infrastructures time to test those changes and deploy those changes. That's potentially many weeks if not months, especially for users at large corporations. I totally understand, of course, that this isn't by any means your problem and hope you don't see my appeal in that light. :] We're just looking for _any_ solution that allows this to be backward compatible in 5.x while we help our users migrate. If this is indeed not at all a possibility, then we'll simply try to get the word out as best we can and rush our packages out the door as quickly as we can do so.

Do you know any more about what the current timing looks like? I'm of course happy to jump on a call or IRC if you'd prefer a private discussion. Cheers and thanks again for being so responsive on this issue. :]

**Comment by DmitryKuzmenko** [comment]
#47106 is a compatibility fix. We have to review all the tornado related code and rework much of it to fully move to the Tornado 5.0 approach. This work is for develop branch. I propose to create a separated issue for this.

**Comment by shish** [comment]
Looks like Debian just released tornado 5 into testing, overwriting tornado 4, so all the salt packages are broken :( I think the only way to get Salt running on Debian Testing is to pip install into a custom virtualenv?

**Comment by amontalban** [comment]
FreeBSD is already using 5.0 so we are getting affected by this.

**Comment by gtmanfred** [comment]
We are close to having tests running for all of this, there are only 2 tests modules failing on 2018.3, and we just need the pyzmq 17.0.0 backported to 2017.7 for that to be ready
